### PR TITLE
Publicize Third Party Authentication methods.

### PR DESCRIPTION
**Pull request by nlutsenko**
Publicize + Update Docs.
Required for opensourcing ParseFacebookUtils, ParseTwitterUtils.

cc @grantland @hallucinogen

### CODE REVIEW

**Comment by hallucinogen** [comment]
Looking awesome! Deferring to @grantland to double-check everything.

File: Parse/PFUser.h [hunk 1]
```diff
@@ -449,6 +449,61 @@ typedef void(^PFUserLogoutResultBlock)(NSError *PF_NULLABLE_S error);
...
+ @note This method shouldn't be invoked directly unless developing a third party authentication provider.
+ @see PFAuthenticationProvider
+
+ @param authenticationProvider The third party authenticaiton provider to be registered.
+ */
++ (void)registerAuthenticationProvider:(id<PFAuthenticationProvider>)authenticationProvider;
+
+/*!
+ @abstract Logs in a user with third party authentication credentials.
+
```
Comment (grantland) [inline Parse/PFUser.h]: @note

File: Parse/PFUser.h [hunk 2]
```diff
@@ -449,6 +449,61 @@ typedef void(^PFUserLogoutResultBlock)(NSError *PF_NULLABLE_S error);
...
+ @param authType The name of the type of third party authenticaiton provider.
+ @param authData The user credentials of the third party authentication provider.
+
+ @returns A `BFTask` that is resolved to `PFUser` when logging in completes.
+ */
++ (BFTask PF_GENERIC(PFUser *) *)logInWithAuthTypeInBackground:(NSString *)authType authData:(NSDictionary *)authData;
+
+/*!
+ @abstract Links this user to a third party authentication provider.
+
```
Comment (grantland) [inline Parse/PFUser.h]: @note

File: Parse/PFUser.h [hunk 3]
```diff
@@ -449,6 +449,61 @@ typedef void(^PFUserLogoutResultBlock)(NSError *PF_NULLABLE_S error);
...
+ @param authType The name of the type of third party authenticaiton provider.
+
+ @returns A `BFTask` that is resolved to `@YES` if unlinking succeeds.
+ */
+- (BFTask PF_GENERIC(NSNumber *) *)unlinkWithAuthTypeInBackground:(NSString *)authType;
+
+/*!
+ @abstract Indicates whether this user is linked with a third party authentication provider of a specific type.
+
```
Comment (grantland) [inline Parse/PFUser.h]: @note

**Comment by grantland** [comment]
LGTM!

File: Parse/PFAuthenticationProvider.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ @returns Returns a unique name for the type of authentication the provider does.
+ */
++ (NSString *)authType;
+
+/*!
+ @abstract Deauthenticates (logs out) the user associated with this provider.
+
+ @returns A task that resolves to anything if deauthentication succeeded, otherwise - faulted task.
+ */
+- (BFTask *)deauthenticateInBackground;
```
Comment (richardjrossiii) [inline Parse/PFAuthenticationProvider.h]: Please genericize these.

Comment (nlutsenko) [inline Parse/PFAuthenticationProvider.h]: I am glad you asked. For both of these - we absolutely don't care about the result type of the task, so I would rather leave this as is.

**Pull request merged**